File: 211-AddandSearchWord/Python/main.py
from collections import defaultdict

class WordDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = defaultdict()
        

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: void
        """
        current = self.root
        for letter in word:
            current = current.setdefault(letter, {})
        
        current["leaf"] = True

    def recsearch(self, word, node):
        if node == True:
            return False
        
        if len(word) == 0:
            return "leaf" in node
        
        res = False
        if word[0] == '.':
            for k in node.keys():
                res = res or self.recsearch(word[1:], node[k])
        elif word[0] in node:
            res = res or self.recsearch(word[1:], node[word[0]])
        
        return res
        
    def search(self, word):
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        return self.recsearch(word, self.root)
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: 211-AddandSearchWord/C++/main.cpp
class WordDictionary {
    struct TrieNode {
        bool word = false;
        unordered_map<char, TrieNode*> children;
        
        explicit TrieNode(bool word=false, unordered_map<char, TrieNode*> children=unordered_map<char, TrieNode*> {}) :
        word(word), children(std::move(children)) {}
    };
    
    std::unique_ptr<TrieNode> root;
public:
    /** Initialize your data structure here. */
    WordDictionary() {
        root = std::make_unique<TrieNode>();
    }
    
    /** Adds a word into the data structure. */
    void addWord(string word) {
        auto current = root.get(); // get raw pointer
        for (auto letter: word) {
            if (current->children.find(letter) == current->children.end()) {
                auto child = new TrieNode;
                current->children.emplace(letter, child);
            }
            current = current->children.at(letter);
        }
        current->word = true;
    }
    
    bool recsearch(string word, TrieNode* node) {               
        if (word.size() == 0) {
            if (node->word == true)
                return true;
            else
                return false;
        }
            
        bool res = false;
        if (word[0] == '.') {
            for (auto key: node->children) {                
                res = res || recsearch(word.substr(1), node->children[key.first]);
            }
        } else if (node->children.find(word[0]) != node->children.end()) {
            res = res || recsearch(word.substr(1), node->children[word[0]]);
        }
        
        return res;
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
        return recsearch(word, root.get());
    }
};


/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * bool param_2 = obj.search(word);
 */